### PR TITLE
Newapi/producer tests

### PR DESCRIPTION
**Pull request by 13h3r**
Tests for new API producer functionality

### CODE REVIEW

File: core/src/test/scala/com/softwaremill/react/kafka2/ProducerTest.scala [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    val result = Mockito.mock(classOf[KafkaProducer[K, V]])
+    Mockito.when(result.send(any[ProducerRecord[K, V]], any[Callback])).thenAnswer(new Answer[java.util.concurrent.Future[RecordMetadata]] {
+      override def answer(invocation: InvocationOnMock) = {
+        val record = invocation.getArguments()(0).asInstanceOf[ProducerRecord[K, V]]
+        val callback = invocation.getArguments()(1).asInstanceOf[Callback]
+        handler(record, callback).onComplete {
+          case Success(value) if !closed => callback.onCompletion(value, null)
+          case Success(value) if closed => callback.onCompletion(null, new Exception("Kafka producer already closed"))
+          case Failure(ex: Exception) => callback.onCompletion(null, ex)
+          case Failure(ex) => ???
```
Comment (kciesielski) [inline core/src/test/scala/com/softwaremill/react/kafka2/ProducerTest.scala]: Is this for the `Throwable` case when Throwable is not an Exception? Maybe this should be called `case Failure(throwableUnsupported)` or something similar to indicate that it's not some unfinished code.

File: core/src/main/scala/com/softwaremill/react/kafka2/Producer.scala [hunk 2]
```diff
@@ -34,32 +34,64 @@ class ProducerSendFlowStage[K, V](producerProvider: () => KafkaProducer[K, V])
...
-        override def onPush(): Unit = {
+        override def onPush() = {
           val msg = grab(messages)
           val result = Promise[(ProducerRecord[K, V], RecordMetadata)]
           producer.send(msg, new Callback {
-            override def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = {
+            override def onCompletion(metadata: RecordMetadata, exception: Exception) = {
               val completion = Option(metadata).map(m => Success((msg, m))).getOrElse(Failure(exception))
               result.complete(completion)
+              decrementConfirmation.invoke(())
```
Comment (13h3r) [inline core/src/main/scala/com/softwaremill/react/kafka2/Producer.scala]: We are talking about kafka's callback, right?
AFAIU callback will be invoked when message will be sent. Callback will be invoked regardless a result of sending - both for success and failure.

File: core/src/test/scala/com/softwaremill/react/kafka2/ProducerTest.scala (hunk 1 above)
Comment (13h3r) [inline core/src/test/scala/com/softwaremill/react/kafka2/ProducerTest.scala]: done

File: core/src/test/scala/com/softwaremill/react/kafka2/ProducerTest.scala [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+  }
+
+  it should "emit confirmation in same order as inputs" in {
+    val input = 1 to 3 map recordAndMetadata
+
+    val client = {
+      val inputMap = input.toMap
+      new ProducerMock[K, V](ProducerMock.handlers.delayedMap(100 millis)(x => Try{ inputMap(x) }))
+    }
+    val probe = Source
```
Comment (13h3r) [inline core/src/test/scala/com/softwaremill/react/kafka2/ProducerTest.scala]: done

File: core/src/main/scala/com/softwaremill/react/kafka2/Producer.scala (hunk 2 above)
Comment (kciesielski) [inline core/src/main/scala/com/softwaremill/react/kafka2/Producer.scala]: Yes, I mean Kafka callback. Just wanted to be sure that it won't be lost (timeout?). If we don't call `decrementConfirmation.invoke()` for some reason, we will end up with incosistency.

Comment (13h3r) [inline core/src/main/scala/com/softwaremill/react/kafka2/Producer.scala]: @kciesielski I scouted a kafka client code and found that the callback and the future completion performed right after the client have a result of a batch sending. If something hang batch sending we will be hanged too. 

Any ideas how to check it?

Comment (kciesielski) [inline core/src/main/scala/com/softwaremill/react/kafka2/Producer.scala]: Could you point me to the sources in Kafka code?

Comment (13h3r) [inline core/src/main/scala/com/softwaremill/react/kafka2/Producer.scala]: In `Sender` there is method `completeBatch` which call `batch.done` which completes Future and call `Callback`

**Comment by 13h3r** [comment]
Anything else I have to do to merge this?

**Comment by kciesielski** [comment]
@13h3r It's ok, thank you. Let's merge.

**Pull request merged**